File: 2025-09/2025-09-27/spamDetector.js
function isSpam(number) {
  const { countryCode, areaCode, localNumber1, localNumber2 } = 
    /\+(?<countryCode>\d+) \((?<areaCode>\d{3})\) (?<localNumber1>\d{3})-(?<localNumber2>\d{4})/
    .exec(number)
    .groups;

  if (countryCode.length > 2 || countryCode[0] !== '0')
    return true;

  if (areaCode < 200 || 900 < areaCode)
    return true;

  const sumOfLocalNumber1 = 
    Array
    .from(localNumber1)
    .reduce((sum, digit) => sum + parseInt(digit), 0)
  if (localNumber2.indexOf(sumOfLocalNumber1) !== -1)
    return true;

  if (/(\d)\1{3,}/.test(countryCode + areaCode + localNumber1 + localNumber2))
    return true

  return false;
}

console.log(isSpam("+0 (200) 234-0182"));
console.log(isSpam("+091 (555) 309-1922"));
console.log(isSpam("+1 (555) 435-4792"));
console.log(isSpam("+0 (955) 234-4364"));
console.log(isSpam("+0 (155) 131-6943"));
console.log(isSpam("+0 (555) 135-0192"));
console.log(isSpam("+0 (555) 564-1987"));
console.log(isSpam("+00 (555) 234-0182"));

// --- TEST SUITE ---

const testsText = `
1. isSpam("+0 (200) 234-0182") should return false.
2. isSpam("+091 (555) 309-1922") should return true.
3. isSpam("+1 (555) 435-4792") should return true.
4. isSpam("+0 (955) 234-4364") should return true.
5. isSpam("+0 (155) 131-6943") should return true.
6. isSpam("+0 (555) 135-0192") should return true.
7. isSpam("+0 (555) 564-1987") should return true.
8. isSpam("+00 (555) 234-0182") should return false.
`;

const testsRegex = /(?<number>\d+)\.\s(?<functionCall>.+) should return (?<output>.+?)\.?$/gm;
const testData = [...testsText.matchAll(testsRegex).map(match => match.groups)];

function runTests(testData) {
  console.log("--------------------------");
  console.log("🧪Starting Verification...");
  console.log("--------------------------");

  let allPassed = true;
  let failCount = 0;
  
  testData.forEach(test => {
    const functionCallOutput = eval(test.functionCall);
    const testOutput = eval(test.output);
    let comparison;

    if (Array.isArray(testOutput))
      comparison = arraysEqual(functionCallOutput, testOutput);
    else
      comparison = functionCallOutput === testOutput;

    if (comparison) {
      console.log(`${test.number}.✅PASS - Function Call: ${test.functionCall}`);
    }
    else {
      console.log(`${test.number}.❌FAIL - Function Call: ${test.functionCall}\nExpected: ${testOutput}\nGot: ${functionCallOutput}`);
      allPassed = false;
      failCount ++;
    }
  })

  console.log("----------------------------");

  if (allPassed)
    console.log("🎉SUCCESS: All tests PASSED.");
  else
    console.log(`⚠️WARNING: ${failCount}/${testData.length} tests FAILED.`);
}

function arraysEqual(a, b) {
  if (a.length !== b.length)
    return false;

  for (const index in a) {
    if (Array.isArray(a[index])) {
      if (Array.isArray(b[index])) {
        if (!arraysEqual(a[index], b[index]))
          return false;
      }
      else 
        return false;
    }

    else if (a[index] !== b[index])
      return false;
  }

  return true;
}

runTests(testData);

File: 2025-09/2025-09-27/spam_detector.py
def is_spam(number):
    import re
    groups_dict = re \
        .match(r"\+(?P<country_code>\d+) \((?P<area_code>\d{3})\) (?P<local_number1>\d{3})-(?P<local_number2>\d{4})", number) \
        .groupdict()

    country_code, area_code, local_number1, local_number2 = (
        groups_dict['country_code'],
        groups_dict['area_code'],
        groups_dict['local_number1'],
        groups_dict['local_number2']
    )

    if len(country_code) > 2 or country_code[0] != "0":
        return True

    if int(area_code) < 200 or 900 < int(area_code):
        return True

    sum_of_local_number1 = sum(map(lambda digit: int(digit), local_number1))
    if str(sum_of_local_number1) in local_number2:
        return True

    if re.search(r"(\d)\1{3,}", country_code + area_code + local_number1 + local_number2):
        return True

    return False

print(is_spam("+0 (200) 234-0182"))
print(is_spam("+091 (555) 309-1922"))
print(is_spam("+1 (555) 435-4792"))
print(is_spam("+0 (955) 234-4364"))
print(is_spam("+0 (155) 131-6943"))
print(is_spam("+0 (555) 135-0192"))
print(is_spam("+0 (555) 564-1987"))
print(is_spam("+00 (555) 234-0182"))

# --- TEST SUITE ---

tests_text = '''
1. is_spam("+0 (200) 234-0182") should return False.
2. is_spam("+091 (555) 309-1922") should return True.
3. is_spam("+1 (555) 435-4792") should return True.
4. is_spam("+0 (955) 234-4364") should return True.
5. is_spam("+0 (155) 131-6943") should return True.
6. is_spam("+0 (555) 135-0192") should return True.
7. is_spam("+0 (555) 564-1987") should return True.
8. is_spam("+00 (555) 234-0182") should return False.
'''

import re

tests_regex = re.compile(r"(?P<number>\d+)\.\s(?P<function_call>.+) should return (?P<output>.+?)\.?$", re.MULTILINE)
test_data = [item.groupdict() for item in re.finditer(tests_regex, tests_text)]

def run_tests(test_data):
    print("--------------------------")
    print("🧪 Starting Verification...")
    print("--------------------------")

    all_passed = True
    fail_count = 0

    for test in test_data:
        function_call_output = eval(test['function_call'])
        test_output = eval(test['output'])
        
        if function_call_output == test_output:
            print(f"{test['number']}.✅ PASS - Function Call:\n{test['function_call']}")
        else:
            print(f"{test['number']}.❌ FAIL - Function Call:\n{test['function_call']}\nExpected:\n{test_output}\nGot:\n{function_call_output}")
            all_passed = False
            fail_count += 1

    print("----------------------------");

    if (all_passed):
        print("🎉 SUCCESS: All tests PASSED.");
    else:
        print(f"⚠️ WARNING: {fail_count}/{len(test_data)} tests FAILED.");

run_tests(test_data)